src/Microsoft.ML.StandardLearners/FactorizationMachine/FactorizationMachine.cs

```diff
@@ -0,0 +1,711 @@
+            {
+                entry.SetProgress(0, iter, _numIterations);
+                entry.SetProgress(1, exampleCount);
+            });
+            Func<int, bool> pred = c => fieldColumnIndexes.Contains(c) || c == data.Schema.Label.Index || (data.Schema.Weight != null && c == data.Schema.Weight.Index);
+            InitializeTrainingState(fieldCount, totalFeatureCount, predictor, out float[] linearWeights,
+                out AlignedArray latentWeightsAligned, out float[] linearAccSqGrads, out AlignedArray latentAccSqGradsAligned);
+            while (iter++ < _numIterations)
+            {
+                using (var cursor = data.Data.GetRowCursor(pred, rng))
+                {
+                    var labelGetter = RowCursorUtils.GetLabelGetter(cursor, data.Schema.Label.Index);
+                    var weightGetter = data.Schema.Weight == null ? null : RowCursorUtils.GetGetterAs<float>(NumberType.R4, cursor, data.Schema.Weight.Index);
+                    for (int i = 0; i < fieldCount; i++)
+                        featureGetters[i] = cursor.GetGetter<VBuffer<float>>(fieldColumnIndexes[i]);
+                    loss = 0;
+                    exampleCount = 0;
+                    badExampleCount = 0;
+                    while (cursor.MoveNext())
+                    {
+                        float label = 0;
+                        float weight = 1;
+                        int count = 0;
+                        float modelResponse = 0;
+                        labelGetter(ref label);
+                        weightGetter?.Invoke(ref weight);
+                        float annihilation = label - label + weight - weight;
+                        if (!FloatUtils.IsFinite(annihilation))
+                        {
+                            badExampleCount++;
+                            continue;
+                        }
+                        if (!FieldAwareFactorizationMachineUtils.LoadOneExampleIntoBuffer(featureGetters, featureBuffer, _norm, ref count,
+                            featureFieldBuffer, featureIndexBuffer, featureValueBuffer))
+                        {
+                            badExampleCount++;
+                            continue;
+                        }
+                        FieldAwareFactorizationMachineInterface.CalculateIntermediateVariables(fieldCount, _latentDimAligned, count,
+                            featureFieldBuffer, featureIndexBuffer, featureValueBuffer, linearWeights, latentWeightsAligned, latentSum, ref modelResponse);
+                        var slope = CalculateLossSlope(label, modelResponse);
+                        FieldAwareFactorizationMachineInterface.CalculateGradientAndUpdate(_lambdaLinear, _lambdaLatent, _learningRate, fieldCount, _latentDimAligned, weight, count,
+                            featureFieldBuffer, featureIndexBuffer, featureValueBuffer, latentSum, slope, linearWeights, latentWeightsAligned, linearAccSqGrads, latentAccSqGradsAligned);
```

Comment:
Similarly, could you add a note here saying that this method implements Algorithm 2 in [3]?